### PR TITLE
add SQS Query API

### PR DESCRIPTION
**Pull request by thrau**
# Summary

This PR adds the [SQS Query API](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-making-api-requests.html), which treats SQS Queue URLs as endpoints to make AWS SQS requests to. This is a prerequisite to make the `asf` SQS provider the default provider.

There are two correcting, but breaking changes related to making requests to the SQS Queue URLs. 

* `ListQueues` and `CreateQueue` are invalid actions to call on a queue url endpoint.
* Making an empty request to the queue url endpoint does not return a valid response, it always returns `<UnknownOperationException/>` with a 404.

# Changes

* SQS endpoint strategy for generating queue URLs like "http://us-east-2.queue.localhost.localstack.cloud/000000000000/my-queue"
* SQS Query API that uses the edge router
* A test suite for the query API
* Some utilities for being able to write those tests
* Some minor parity fixes

## Endpoint strategy (how SQS Queue URLs are formatted)

A new feature that is necessary to make the query API work for multi-region is the `SQS_ENDPOINT_STRATEGY`, similar to what we do in [opensearch](https://docs.localstack.cloud/aws/opensearch/#endpoints).
The endpoint strategies and their corresponding queue URLs are:
* `off`: `http://localhost:4566/000000000000/MyQueue` (currently the default, note how we cannot infer a region other than the default region. needed for backwards compatibility)
* `path`: `http://localhost:4566/queue/eu-central-1/000000000000/MyQueue` (needed for when `localhost.localstack.cloud` does not resolve
* `domain`: `http://us-east-2.queue.localhost.localstack.cloud:4566/000000000000/MyQueue` (the default, mimics [Amazon SQS legacy endpoints](https://docs.aws.amazon.com/general/latest/gr/sqs-service.html)

This would need documentation @HarshCasper 

## SQS Query API

The query API is an HTTP request handler that acts as a wrapper around the active SQS provider.
It works as follows:

* Expose queue URLs as HTTP routes on the edge proxy
* Translate those into AWS requests based on the Queue URL
* Perform the request via boto

To that end we inject for each endpoint strategy a handler into the `edge.ROUTER` that takes precedence over the service routes.

## Test suite for the query API

I added a comprehensive test suite for the query API that is validated against AWS and demonstrates how some of the previous behavior was diverging from AWS (see summary).

To implement the test suite, I added utilities that make it easier to sign of AWS HTTP requests using botocore's `BaseSigner` abstraction. `SigningHttpClient` the `aws_http_client_factory` creates instances of these clients for a particular service using the credentials from a botocore session (this allows us to use it together with `TEST_TARGET = AWS_CLOUD`.
Please take a look at this @dominikschubert

## Other changes

Along the way I fixed a few related issues

* Parity for `InvalidAttributeName` messages
* Parity for `QueueDoesNotExist` messages
* Cleaned up `test_get_list_queues_with_auth_in_presigned_url` to use the query-style authparams that the new `SigningHttpClient` provides

# Background

Currently, we _can_ perform AWS requests directly on the SQS URLs (although it does not behave like AWS). With the `ProxyListener` model, implementing this type of behavior is fairly straight forward:

* First, we associate requests to the URL with the SQS service: https://github.com/localstack/localstack/blob/81f8f56fc88acdda93e8b47b574d189776cb571e/localstack/services/edge.py#L435-L437
* Then we modify the HTTP request to look like an AWS request: https://github.com/localstack/localstack/blob/81f8f56fc88acdda93e8b47b574d189776cb571e/localstack/services/sqs/sqs_listener.py#L294-L299

## Parity problems

There was behavior introduced in #2622 that translates an empty GET request to the queue URL into `GetQueueUrl` request, which is not how AWS behaves.

### Empty GET request on queue URLs

* AWS Response:
  ```
  % curl "https://queue.amazonaws.com/878966065785/MyQueue"                   
  <UnknownOperationException/>
  ```
* Current LocalStack Response:
  ```
  % curl "http://localhost:4566/000000000000/MyQueue"
  <GetQueueUrlResponse><GetQueueUrlResult><QueueUrl>http://localhost:4566/000000000000/MyQueue</QueueUrl></GetQueueUrlResult><ResponseMetadata><RequestId>S86VYS8ILUZE7BEIJL87E9S4R09YR52YUKHWJINFZTOPAY7AFEGB</RequestId></ResponseMetadata></GetQueueUrlResponse>
  ```

### Sending invalid requests

Trying to send an (invalid) `SendMessage` request:
```
 % curl "http://localhost:4566/000000000000/MyQueue?Action=SendMessage"                       
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<title>500 Internal Server Error</title>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.</p>
```

### Using `application/json` accept type

Using JSON (should return a json dict, which should include the attributes)
```
 % curl -H "Accept: application/json" "http://localhost:4566/000000000000/MyQueue?Action=GetQueueAttributes&AttributeName.1=All"
<GetQueueAttributesResponse>...
```

# TODO

- [x] basic implementation
- [x] rebase after merging #5927
- [x] write parity tests against AWS (involves creating [AUTHPARAMS](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-making-api-requests.html) for each request)
- [x] improve current implementation to cover at least what we were able to do so far in the current provider
- [x] decide on whether to fix parity (empty request => returns UnknownOperationException) and introduce a breaking localstack change
- [x] ~consider implementing JSON serializing (nice to have)~

### CODE REVIEW

**Comment by thrau** [comment]
I'd like to get your opinion on this @baermat and @alexrashed.
I'm not confident whether this is the best solution. We're essentially wrapping the AWS API, but using parser/serializer internals which is the very thing that underlies the AWS API. There may be some easier or more clever way of doing it. It's also not 100% clear to me whether Query API is something special, or just a subset of operations that require the Queue URL, or whether it's essentially a special SQS API spec, where the QueueURL parameter is encoded in the URL. Maybe it would be easier to implement this on the parser level :shrug:

**Comment by alexrashed** [comment]
While the implementation of this feature looks good (I would only have some minor suggestions), I think we should tackle this problem differently, as you already indicated in your comment above:

- In my opinion, the router should only be used for routes which should not go through ASF (like localstack internal routes, or dynamically spawned routes like API Gateway routes).
- I would tackle this issue here with a service specific parser responsible for handling such service-specific requests (i.e. handling the queue URLs and extracting params like the QueueName from there).
- In addition, a service-specific extension to the service router (which is currently being implemented in https://github.com/localstack/localstack/pull/5734) would be necessary to properly encapsulate this feature. However, such an extension mechanism is not in the scope of this PR but should be enhanced once it's merged. The initial implementation will preserve the current logic (i.e. the custom rule for query URLs for SQS), which means initially just the service specific parser would need to be adjusted / implemented.

**Comment by thrau** [comment]
I decided to go with the current implementation over implementing it as a parser. The query has a few peculiarities over the default services that would probably be very awkward to handle transparently by modifying the request.

I also think we should break the current behavior of the queue url (which seems completely arbitrary and has no foundation in the AWS API behavior)

**Comment by thrau** [comment]
I'll leave the JSON behavior as a future feature as that seems to be part of the serialzier and therefore out of scope of this PR

**Comment by thrau** [comment]
@viren-nadkarni just a heads up, this will create lots of conflicts with #6016 and i will help rebase the PR :-)

File: localstack/services/sqs/query_api.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    """Uses the endpoint host to extract the region. See:
+    https://docs.aws.amazon.com/general/latest/gr/sqs-service.html"""
+    return handle_request(request, region)
+
+
+@route(
+    '/<regex("[0-9]{12}"):account_id>/<regex("[a-zA-Z0-9_-]+(.fifo)?"):queue_name>',
+    methods=["POST", "GET"],
+)
+def legacy_handler(request: Request, account_id, queue_name) -> Response:
```
Comment (baermat) [inline localstack/services/sqs/query_api.py:51]: The legacy handler has return type -> Response, the domain and path one do not. Just checking if that is intentional

File: localstack/services/sqs/query_api.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        raise UnknownOperationException()
+
+    if action in ["ListQueues", "CreateQueue"]:
+        raise InvalidAction(action)
+
+    # prepare aws request
+    params = {"QueueUrl": request.base_url}
+    # if a QueueUrl is already set in the body, it should overwrite the one in the URL
+    params.update(request.values)
+    body = urlencode(params)
```
Comment (baermat) [inline localstack/services/sqs/query_api.py:119]: Is that documented somewhere/tested against AWS? I guess the POST tests would cover this once implemented. Maybe put a TODO there for now if it is an assumption.

File: localstack/services/sqs/query_api.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+    try:
+        operation, service_request = parser.parse(
+            Request("POST", "/", headers=request.headers, body=body)
+        )
+    except OperationNotFoundParserError:
+        raise InvalidAction(action)
```
Comment (baermat) [inline localstack/services/sqs/query_api.py:126]: I find it interesting that the parser always takes a POST request, regardless of what the original call was. I am assuming that is intentional, just making sure :)

File: localstack/services/sqs/provider.py [hunk 4]
```diff
@@ -267,10 +268,24 @@ def arn(self) -> str:
...
+        or based on the 'Host' request header"""
+
         host_url = context.request.host_url
-        if config.SQS_PORT_EXTERNAL:
-            host_url = external_service_url("sqs")
+
+        if config.SQS_ENDPOINT_STRATEGY == "domain":
+            # queue.localhost.localstack.cloud:4566/000000000000/my-queue (us-east-1)
+            # or us-east-2.queue.localhost.localstack.cloud:4566/000000000000/my-queue
+            region = "" if self.key.region == "us-east-1" else self.key.region + "."
```
Comment (baermat) [inline localstack/services/sqs/provider.py:279]: This is some interesting AWS behaviour^^

File: tests/integration/test_sqs.py [hunk 5]
```diff
@@ -1637,21 +1638,338 @@ def test_call_fifo_queue_url(self, sqs_client, sqs_create_queue):
...
+        )
+        assert response.ok
+        assert "<DeleteQueueResponse " in response.text
+
+        def queue_is_removed():
+            try:
+                result = sqs_client.get_queue_url(QueueName=queue_url.split("/")[-1])
+                return result.get("QueueUrl") != queue_url
+            except ClientError:
+                return True
```
Comment (baermat) [inline tests/integration/test_sqs.py:1757]: Is it safe here to return True on any ClientError, or does this have the potential to be green on arbitrary issues?

File: tests/integration/test_sqs.py [hunk 6]
```diff
@@ -1637,21 +1638,338 @@ def test_call_fifo_queue_url(self, sqs_client, sqs_create_queue):
...
+        queue_url = sqs_create_queue()
+
+        sqs_client.delete_queue(QueueUrl=queue_url)
+
+        def queue_is_removed():
+            try:
+                result = sqs_client.get_queue_url(QueueName=queue_url.split("/")[-1])
+                return result.get("QueueUrl") != queue_url
+            except ClientError:
+                return True
```
Comment (baermat) [inline tests/integration/test_sqs.py:1812]: Same as before. Might hide the true error source here

**Review by baermat — COMMENTED** [review]
Some nitpicks but otherwise LGTM!
Amazing that you cleaned up those test cases where the request was generated from scratch!

File: localstack/services/sqs/query_api.py (hunk 3 above)
Comment (thrau) [reply localstack/services/sqs/query_api.py:126]: yes actually, great catch! sqs uses the AWS query protocol, and in the query protocol each operation is a POST request whether the request is encoded as a query string in the request body. it's just how the protocol is specified. with the query API, we can make GET requests, so we translate them to POST requests to the backend API.

File: localstack/services/sqs/provider.py (hunk 4 above)
Comment (thrau) [reply localstack/services/sqs/provider.py:279]: true not very clean! it's just how AWS behaves.

so it's either "queue.amazonaws.com" for us-east-1 queues, and "<region>.queue.amazonaws.com" for all others. :shrug:

File: tests/integration/test_sqs.py (hunk 5 above)
Comment (thrau) [reply tests/integration/test_sqs.py:1757]: that's true, we should probably check whether it's a non-existent queue exception, good catch!

File: localstack/services/sqs/query_api.py (hunk 1 above)
Comment (thrau) [reply localstack/services/sqs/query_api.py:51]: lazyness! will fix

File: localstack/services/sqs/query_api.py (hunk 2 above)
Comment (thrau) [reply localstack/services/sqs/query_api.py:119]: yes that's tested against AWS. i'll add a comment

**Pull request merged**